gui/src/containers/JobSpecRuns.js

```diff
@@ -23,10 +25,54 @@ const styles = theme => ({
   title: {
     marginTop: theme.spacing.unit * 5,
     marginBottom: theme.spacing.unit * 5
+  },
+  customButtons: {
+    flexShrink: 0,
+    color: theme.palette.text.secondary,
+    marginLeft: theme.spacing.unit * 2.5
   }
 })
 
-const START_PAGE = 1
+const TableButtons = props => {
+  const lastPage = Math.ceil(props.count / props.rowsPerPage) - 1
+  const currentPage = props.page - 1
+  const handleFirstPage = e => {
+    if(props.history)
+      props.history.replace(`/job_specs/${props.specID}/runs/page/1`)
+    props.onChangePage(e, 0)
+  }
+  const handleLastPage = e => {
+    if(props.history)
+      props.history.replace(`/job_specs/${props.specID}/runs/page/${lastPage + 1}`)
+    props.onChangePage(e, Math.max(0, lastPage))
+  }
+  const handlePrevPage = e => {
+    if(props.history)
+      props.history.replace(`/job_specs/${props.specID}/runs/page/${currentPage}`)
+    props.onChangePage(e, currentPage - 1)
+  }
+  const handleNextPage = e => {
+    if(props.history)
+      props.history.replace(`/job_specs/${props.specID}/runs/page/${currentPage + 2}`)
+    props.onChangePage(e, currentPage + 1)
```

Comment:
These methods seem a bit chaotic. We're juggling 0 based and 1 based page numbers right next to each other, checking `props.history` multiple times, etc.

Perhaps we can DRY this up with some currying like so (assuming 1 based indexed pages):

```js
const handlePage = page => {
  page = Math.min(page, lastPage)
  page = Math.max(page, firstPage)
  const curry = e => {
    if(props.history) {
      props.history.replace(`/job_specs/${props.specID}/runs/page/${page}`)
    }
    props.onChangePage(e, page)
  }
  return curry
}
...

<IconButton onClick={handlePage(0)} disabled={currentPage === 0} />
<IconButton onClick={handlePage(currentPage+1)} disabled={currentPage === 0} />
```

You'll want to double check this, but I hope the concept is sound.